### DISCUSSION TITLE
Annotation IDs and dates

### DISCUSSION THREAD

**Discussion by jesuslop** [sumatrapdfreader/sumatrapdf#4544] [General]
Hi, I'm working in an annotation synchronization tool, and wanted to request that PDF annotations be created with an ID (so it is baked to an "/NM" entry in the annotation object at the PDF level), Now it isn't filled (just tried for a highligh, Sumatra 3.5.2)

Also I would like that when an already existing annotation is changed (as in changing the color of a highlight, or the text of a comment note), that the modification date be updated accordingly. ("/M" PDF object property). Now the modification date is set to the creation date, and does not change.

Not sure if its of help, in pymupdf there is some fiddling to do to add annotation IDs at the annotation level of the API, but it's doable at a lower level of manipulating objects by "xref". Think sumatra depends on mupdf, as does pymupdf, so here goes my recipe to fill annotation IDs for whole document, in a uncertain hope to help/inspire.

```
PDF_NAME_KEY = "NM"

def force_annotation_id(doc, xref, new_id):
	doc.xref_set_key(xref, PDF_NAME_KEY, pymupdf.get_pdf_str(new_id))
 
# see https://pymupdf.readthedocs.io/en/latest/page.html#Page.annots
def fix_annotation_ids(pdf_doc):
	for page in pdf_doc.pages():
		xrefs = [annot.xref for annot in page.annots(types=ANNOT_TYPES)]
		for xref in xrefs:
			if not PDF_NAME_KEY in pdf_doc.xref_get_keys(xref):
				new_id = "xxx:page-{0}:{1}".format(page.number, uuid.uuid4())
				force_annotation_id(pdf_doc, xref, new_id)
```

**Comment by GitHubRulesOK** [comment]
MuPDF does not generate "Document Fingerprinting" (the ID in the source file and others when added to.) 
Thus nothing for SumatraPDF to inherit there from MuPDF behaviours.
the date when added is naturally the time the file was modified `/M(D:20240922003505Z)` That entry is all that is needed for the annotation to work.

Here is a file totally generated by MuDraw from a Windows native Write DocX
I have left the body in "Plain Text" for simple inspection in Notepad or Write
[MuHelloWorld.pdf](https://github.com/user-attachments/files/17087837/MuHelloWorld.pdf)

```
xref
0 12
0000000000 00001 f 
                ...
0000071118 00000 n 

trailer
<</Size 12/Root 1 0 R>>
startxref
71226
%%EOF
                ...
12 0 obj
<</Type/Annot/Subtype/FreeText/Rect[100.5 450.73643 300.5 550.7364]/BS<</Type/Border/W 1>>/DA(/Helv 12 Tf 0 0 0 rg)/P 11 0 R/F 4/M(D:20240922003505Z)/T(K)/AP<</N 14 0 R>>/Contents(From MuPDF)/BE<</S/C/I 1>>>>
endobj
                ...
xref
11 4
0000071532 00000 n 
0000071655 00000 n 
0000071881 00000 n 
0000071971 00000 n 

trailer
<</Size 15/Root 1 0 R/Prev 71226>>
startxref
72261
%%EOF
```


So MuPDF Trailer in source data and annotations is kept light and basically fairly non traceable (apart from the MuPDF signature).

MS Edge (Powered by Adobe Acrobat) will NOT add an ID unless it has to, but does use 2 different Date entries for its /M(od)'s
`/M(D:20240922022524+01'00')`
[EdgeHelloWorld.pdf](https://github.com/user-attachments/files/17087896/EdgeHelloWorld.pdf)

and bloat
```
18 0 obj
<</ModDate(D:20240922022530+01'00')>>
endobj
xref
1 1
0000072417 00000 n
12 1
0000072538 00000 n
15 4
0000072814 00000 n
0000072904 00000 n
0000074097 00000 n
0000074458 00000 n
trailer
<</Size 19/Root 1 0 R/Prev 72261>>
startxref
74512
%%EOF
```

For trackability use "Editors"
```
					<rdf:li xml:lang="x-default">HelloWorld.docx</rdf:li>
				</rdf:Alt>
			</dc:title>
			<dc:format>application/pdf</dc:format>
			<xmp:CreateDate>2024-09-22T01:00+01:00</xmp:CreateDate>
			<xmp:ModifyDate>2024-09-22T03:13:17+01:00</xmp:ModifyDate>
			<pdf:Producer>Microsoft: Print To PDF</pdf:Producer>
			<xmpMM:DocumentID>uuid:d6371a6a-aec8-498c-b775-a6cb7300434b</xmpMM:DocumentID>
			<xmpMM:InstanceID>uuid:7c2db4f8-fe31-47ff-a737-dc7a20e8e9ef</xmpMM:InstanceID>

         ...
13 0 obj
<</AP<</N 23 0 R>>/BS<</W 0>>/Contents(From Tracker)/CreationDate(D:20240922030828+01'00')/DA(0 0 0 rg /F2 12 Tf)/DS(font-family:Arial;font-size:12pt;color:#000000;)/F 4/IT/FreeTextTypewriter/M(D:20240922031215+01'00')/NM(a995c105-003a-4ef1-8b5a9dd25beecf9d)/P 9 0 R/Rect[174.71535 509.62373 246.87134 523.0277]/Subj(FreeType)/Subtype/FreeText/Type/Annot>>
endobj
      ...
```
[MSPrintHelloWorld-FromTracker.pdf](https://github.com/user-attachments/files/17088013/MSPrintHelloWorld-FromTracker.pdf)

**Reply by jesuslop** [reply]
Hi! thanks for the prompt reply! Let's check if I understand. I requested that annotations had IDs (in my case for creating annotation synchronization tooling) but the objection is on privacy grounds. Surprisingly to me, this worry is sound, I read that UUIDs can pin you up, and recall something about the MAC being in the recipe. Fine point. But what about using just a pure pseudorandom ID? If it is felt as controversial (?), It could be made configurable in advanced settings. No problem on my side if defaulted to not create IDs. This is thinking in the user community, in my case I already have code to work-around having no IDs. 

As to the annotations modification date, I don't think I'm suggesting something crazy. The PDF spec 1.7, also 2, annotations chapter, defines "M" thus: "The date and time when the annotation was most recently modified." This will be in general different of the creation or modification time of the pdf file or annotation creation time. Has to change in updates (to be compliant).

Why do I want this? I study and work with academic papers and do it in a variety of devices, including my laptop with Sumatra (my kudos by the way), and other closed firmware devices as big screen e-ink readers, and more. And I want to have a coherent study/annotation/review user experience in all of the devices, in a way that if I'm working in, say, Dijkstra 1959, I don't have to care what screen I'm on. I want a magic that keeps all underlying copies updated. File sync is not granular enough (there's need of an n-way merge _a la_ git), and at annotation level if annotations have no date of themselves, I cannot sort their updates in time, among multiple pdf working copies of the same paper, and cannot say which is the latest, definitive change.

**Reply by GitHubRulesOK** [reply]
The "Writing" time of an addition **or deletion** is supposedly **no less** than the copyright year /M(D:1234) the time of document change i.e. /M there is no valid point in writing that twice, per entry. When modified it **may** have a new time. (Acrobat in Edge WILL do that but, the original will still be untouched with its own historic time) This is a source of MetaData Leakage but inherent in PDF. Personally I see no point in altering the time or date as it is non essential to the task of identifying there was a change, I.E. there are 2 entries one later addition but when is immaterial, although Edge then adds a second date moniker.
MuPdf will simply alter the visual changes and leave the source date as it is.
```
15 0 obj
<</Type/Annot/Subtype/Text/Rect[176.3474 500.19264 194.3474 518.1926]/C[1 1 0]/Popup 16 0 R/P 3 0 R/F 28/M(D:20240922200714Z)/T(K)/AP<</N 17 0 R>>/Contents(written)>>
endobj
15 0 obj
<</Type/Annot/Subtype/Text/Rect[176.3474 500.19264 194.3474 518.1926]/C[1 1 0]/Popup 16 0 R/P 3 0 R/F 28/M(D:20240922200714Z)/T(K)/AP<</N 18 0 R>>/Contents(modified)/Name/Paragraph>>
endobj
15 0 obj
<</Type/Annot/Subtype/Text/Rect[207.85051 516.5888 225.85051 534.5888]/C[0 0 1]/Popup 16 0 R/P 3 0 R/F 28/M(D:20240922200714Z)/T(K)/AP<</N 19 0 R>>/Contents(modified)/Name/Help>>
endobj
```

**NOTE: this behaviour is NOT SumatraPDF** 
---
It is the underlying MuPDF decides what is essential for meeting the ISO standards. So when edited by MuPDF it is that behaviour you see
```
15 0 obj
<</Type/Annot/Subtype/Text/Rect[227.36307 491.84629 245.36307 509.84629]/C[.5019608 0 0]/Popup 16 0 R/P 3 0 R/F 28/M(D:20240922200714Z)/T(K)/AP<</N 20 0 R>>/Contents(modified again)/Name/Key>>
endobj
```

RE Privacy there have frequently been complaints about unnecessary MetaData leakage and as I noted Microsoft were previously heavily criticised for embedding User Data in office documents (I have used that in the past for tracking who made document edits for "several means" but not sanctioned as such by those authors).

Microsoft+Acrobat plug-in do NOT add IDs by default in PDF modifications only in the source generation XMP !
